phi2/model.py

```diff
@@ -0,0 +1,185 @@
+from typing import Optional
+from dataclasses import dataclass
+from mlx.utils import tree_unflatten, tree_map
+from transformers import AutoTokenizer
+
+import mlx.core as mx
+import mlx.nn as nn
+import math
+
+@dataclass
+class ModelArgs:
+    max_sequence_length: int = 2048
+    num_vocab: int = 51200
+    model_dim: int = 2560
+    num_heads: int = 32
+    num_layers: int = 32
+    rotary_dim: int = 32
+
+
+class RoPEAttention(nn.Module):
+    def __init__(self, dims: int, num_heads: int, bias: bool = True):
+        super().__init__()
+
+        self.num_heads = num_heads
+
+        self.rope = nn.RoPE(dims // num_heads, traditional=True)
```

Comment:
Shouldn't this be 32 (`rotary_dim`)?.

Compare to candle's implementation: https://github.com/huggingface/candle/pull/1429/files#diff-d3c3d063548726529c19aa36e11195f0f0cfbc07eadcf4e9662a5841086c1d7eR68

Reply:
It seems to work with `self.rope = nn.RoPE(32, traditional=False)`

```
$ python model.py

Answer: Mathematics can be compared to a lighthouse, guiding us through the vast ocean of knowledge. Just as a lighthouse illuminates the path for ships, mathematics illuminates our understanding of the world. It provides a solid foundation upon which we can build our knowledge and navigate through complex problems.

Exercise 2:
Exercise: Explain the concept of logic using the analogy of a puzzle.
```

Reply:
Ohhh that looks promising :)

Reply:
Yup, that solves it! Here are the new values after RoPE, which align with the HF values:

```
[[-0.8564 -2.682  -2.521   3.49    1.593 ]
 [ 0.591  -0.7925 -0.669   2.037   1.59  ]
 [ 4.906  -3.598  -1.939   4.543   2.436 ]
 [ 0.67   -0.6597 -1.07    2.781   1.979 ]
 [-2.322   0.295   1.13    4.168   2.352 ]
 [-4.27    4.164   2.84    3.617   2.467 ]
 [ 0.1169  0.5737  1.362   2.162   1.313 ]
 [ 3.46    3.234   5.24    3.672   3.707 ]
 [ 0.663   0.907   2.41    1.901   3.16  ]
 [ 0.2866  1.38    1.626   0.532   1.244 ]
 [-3.293  -0.743   1.906   1.316   3.875 ]
 [-0.633  -0.4885  0.8003  0.559   3.277 ]
 [-1.244  -1.572   1.029  -1.902   0.2166]
 [ 0.917  -2.59   -1.603  -0.1338  0.0725]
 [ 1.44    0.891  -1.869  -0.924   0.0415]
 [ 0.5444  0.5874  0.386  -2.371  -0.5903]
 [-0.61    1.15   -1.463  -2.025   0.1357]
 [-4.098   4.04   -3.691  -3.98   -0.592 ]
 [ 0.4214  1.346  -2.066  -2.045  -0.0752]
 [ 0.5938  0.1362 -2.654  -2.49   -0.206 ]
 [ 3.541   1.35   -1.801  -5.16    2.088 ]
 [ 0.6724  0.1377 -0.751  -2.848   1.688 ]
 [ 0.1547 -0.803  -0.2246 -2.297  -0.2339]]
```

Reply:
Thanks @jbochi and @jbarrow let's clean this up and land it! @jbarrow should I do some work there or are you on it?

Reply:
I am trying to get fp16 to work as well, just FYI

Reply:
I can only work on it in fp16, haha. (see attached screenshot)

I won't have time for another 6-7 hours to work on it, but am happy to do it then. Otherwise, feel free to go for it.

<img width="399" alt="Memory pressure running phi-2 in fp16" src="https://github.com/ml-explore/mlx-examples/assets/942625/74d4c3fe-fcb2-4959-8561-aafa416557c1">

Reply:
But one last thing, when I run the full model forward -- in fp16 -- I now get roughly the same values as the full HF model -- in fp16, on a T4 -- for all inputs, up to ~0.01 tolerance. So very happy with that.

MLX
```
array([[[6.46128, 5.09739, 1.96567, ..., -2.90429, -2.90344, -2.90395],
        [5.22273, 6.44388, 5.19575, ..., -0.169703, -0.170586, -0.171148],
        [4.31223, 5.27566, 5.94652, ..., -0.375012, -0.375973, -0.375772],
        ...,
        [10.1819, 11.4079, 12.613, ..., 0.708903, 0.708031, 0.706881],
        [10.7088, 12.0519, 13.2024, ..., 1.62455, 1.6232, 1.62217],
        [10.7146, 11.9357, 12.3973, ..., 0.22852, 0.227845, 0.22809]]]
```

HF:
```
tensor([[[ 6.4688,  5.1016,  1.9658,  ..., -2.9043, -2.9043, -2.9043],
         [ 5.2188,  6.4414,  5.1914,  ..., -0.1852, -0.1862, -0.1866],
         [ 4.3516,  5.3281,  5.9922,  ..., -0.3689, -0.3699, -0.3696],
         ...,
         [10.4141, 11.7031, 12.5859,  ...,  0.7778,  0.7769,  0.7754],
         [10.7188, 11.7891, 13.3125,  ...,  1.6123,  1.6113,  1.6104],
         [10.8047, 12.0234, 12.4375,  ...,  0.2321,  0.2314,  0.2317]]],
```

Reply:
Interesting.. when I run it in fp16 it generates bogus text. Not sure why yet

Reply:
Oh, same with the cleaned up version (I get `nan`'s in the logits for some reason).

Reply:
That is strange.. I wonder if it's implicitly upcasting somewhere?

Reply:
It could also be related to not using the mask... adding the mask causes issues in lower precision because you get a sequence of `inf` in the softmax which turns into a NaNs and then you are hosed..

Reply:
I just pushed a fix to get it to generate with fp16 weights. Can you try and lmk if it works for you?

Reply:
I can confirm it works with fp16 on my M2 Pro. 🥳   I was getting garbage too before your fix.

Reply:
Works on an M2 Air!

<img width="867" alt="Screenshot 2023-12-14 at 11 27 04 AM" src="https://github.com/ml-explore/mlx-examples/assets/942625/81975ee0-158f-4d57-8c73-1e6110a9dbd3">